homeassistant/components/homewizard/sensor.py

```diff
@@ -452,3 +519,74 @@ def native_value(self) -> StateType:
     def available(self) -> bool:
         """Return availability of meter."""
         return super().available and self.native_value is not None
+
+
+class HomeWizardExternalSensorEntity(HomeWizardEntity, SensorEntity):
+    """Representation of externally connected HomeWizard Sensor."""
+
+    def __init__(
+        self,
+        coordinator: HWEnergyDeviceUpdateCoordinator,
+        description: HomeWizardExternalSensorEntityDescription,
+        device_unique_id: str,
+    ) -> None:
+        """Initialize Externally connected HomeWizard Sensors."""
+        super().__init__(coordinator)
+        self.entity_description = description
+        self._device_id = device_unique_id
+        self._suggested_device_class = description.suggested_device_class
+        self._attr_unique_id = f"{DOMAIN}_{device_unique_id}"
+        self._attr_device_info = DeviceInfo(
+            identifiers={(DOMAIN, device_unique_id)},
+            name=description.device_name,
+            manufacturer="HomeWizard",
+            model=coordinator.data.device.product_type,
+            serial_number=device_unique_id,
+        )
+        if coordinator.data.device.serial is not None:
+            self._attr_device_info[ATTR_VIA_DEVICE] = (
+                DOMAIN,
+                coordinator.data.device.serial,
+            )
+
+    @property
+    def native_value(self) -> float | int | str | None:
+        """Return the sensor value."""
+        return self.device.value if self.device is not None else None
+
+    @property
+    def device(self) -> ExternalDevice | None:
+        """Return ExternalDevice object."""
+        return (
+            self.coordinator.data.data.external_devices[self._device_id]
+            if self.coordinator.data.data.external_devices is not None
+            else None
+        )
+
+    @property
+    def available(self) -> bool:
+        """Return availability of meter."""
+        return super().available and self.device is not None
+
+    @property
+    def native_unit_of_measurement(self) -> str | None:
+        """Return unit of measurement based on device unit."""
+        if (device := self.device) is None:
+            return None
+
+        # API returns 'm3' but we expect m³
+        if device.unit == "m3":
+            return UnitOfVolume.CUBIC_METERS
+
+        return device.unit
+
+    @property
+    def device_class(self) -> SensorDeviceClass | None:
+        """Validate unit of measurement and set device class."""
+        if (
+            self.native_unit_of_measurement
+            not in DEVICE_CLASS_UNITS[self._suggested_device_class]
```

Comment:
I'm not sold on this approch. Is there really that many different possible native units of this device per suggested device class that we can't create entity descriptions for each combination? Then we could handle special cases like m3 above already in the mapping.

Reply:
It is up to the device manufacturer which unit to use and how to format that. Nothing about that is specified in any standard AFAIK. The current method supports all combination that are allowed by Home Assistant.

When we specify each combination we have to collect those combinations. That is absolutely an option. In the beginning we will see some issues pop up but when most combinations are added it works the way we hope. But I think the end result is less 'clean' as it is now.

Reply:
Even if we would include all allowed native units in the list of combinations it would only result in 23 items currently. That's just 18 more than what we have now. If it would become more than a couple of hundred combinations I'd understand that we don't want to list all of them.

It's always easier to read a declarative approach than a calculated approach.